File: src/swagger/Operation.php
<?php

namespace Ething\Swagger;



class Operation {
	
	private $data;
	private $path;
	private $method;
	
	public function __construct($method, $data, $path) {
		$this->data = $data;
		$this->path = $path;
		$this->method = $method;
		
		if(!isset($this->data->operationId)){
		
			// generate a unique operation id
			
			$opId = $this->method . '_' . $this->path->path();
			
			$opId = preg_replace('/[\s!@#$%^&*()_+=\[{\]};:<>|.\/?,\'"-]/', '_', $opId);
			$opId = preg_replace('/((_){2,})/', '_', $opId);
			$opId = preg_replace('/^_*/', '', $opId);
			$opId = preg_replace('/_*$/', '', $opId);
			
			$this->data->operationId = $opId;
		}
	}
	
	
	public function name() {
		return $this->data->operationId;
	}
	
	
	public function method() {
		return $this->method;
	}
	
	public function description() {
		return empty($this->data->description) ? '' : $this->data->description;
	}
	
	public function path() {
		return $this->path;
	}
	
	public function swagger() {
		return $this->path->swagger();
	}
	
	public function produces() {
		$produces = isset($this->data->produces) ? $this->data->produces : $this->swagger()->produces();
		return $produces;
	}
	
	public function consumes() {
		$comsumes = isset($this->data->consumes) ? $this->data->consumes : $this->swagger()->consumes();
		if(empty($comsumes))
			$comsumes = array('application/json'); // default to json
		return $comsumes;
	}
	
	public function parameters() {
		
		if(!isset($this->parameters_)){
			$parameters = $this->path->parameters();
			
			if(!empty($this->data->parameters)) {
				foreach($this->data->parameters as $param){
					
					if(isset($param->{'$ref'})){
						// resolve the reference
						$param = $this->swagger()->getRef($param->{'$ref'});
					}
					
					$parameters[] = $param->in == 'body' ? new BodyParameter($this, $param) : new StandardParameter($this, $param);
				}
			}
			
			$this->parameters_ = $parameters;
		}
		
		return $this->parameters_;
	}
	
	
	
	public function createRequest(array $data = null, array $options = array()) {
		
		// build the url
		$schemes = isset($this->data->schemes) ? $this->data->schemes : $this->swagger()->schemes();
		if(empty($schemes))
			$schemes = array('http');
		$scheme = (!empty($options['scheme']) && in_array($options['scheme'], $schemes)) ? $options['scheme'] : $schemes[0];
		
		$url = $scheme .'://' . $this->swagger()->host() .  $this->swagger()->basePath(). $this->path->path();
		
		$method = $this->method;
		$query = array();
		$formData = array();
		$headers = array();
		$body = null;
		$hasFiles = false;
		$hasContentTypeHeader = false;
		$hasBody = false;
		
		$consumes = $this->consumes();
		
		$contentType = !empty($options['requestContentType']) ? $options['requestContentType'] : (in_array('application/json', $consumes) ? 'application/json' : $consumes[0]);
		
		
		foreach( $this->parameters() as $param){
			
			$hasValue = isset($data) && isset($data[$param->name()]);
			
			$hasDefault = ($param instanceof StandardParameter) && $param->hasDefault();
			
			if($hasValue){
				$value = $data[$param->name()];
			}
			else {
				if($param->isRequired()){
					if($hasDefault){
						$value = $param->geDefaultValue();
					} else {
						throw new Exception("missing required param: '{$param->name()}'");
					}
				}
				else
					continue;
			}
			
			// validate the incomming data
			if(!$param->validate($value))
				throw new Exception("invalid param: '{$param->name()}'");
			
			
			$in = $param->in();
			switch($in){
				
				case 'formData' : 
					
					if($param->type() == 'file'){
						if(!isset($finfo))
							$finfo = new \finfo(FILEINFO_MIME);
						$formData[$param->name()] = array(
							"type" => "file",
							"content" => $value,
							"filename" => 'file'.count($files),
							"mime" => $finfo->buffer($value)
						);
						
						$hasFiles = true;
					} else {
						$formData[$param->name()] = array(
							"type" => "text",
							"content" => $value
						);
					}
					
					break;
					
				case 'query' : 
					
					$el = '';
					if(is_array($value) && $param->collectionFormat() == 'multi'){
						if($param->collectionFormat() == 'multi'){
							$p = array();
							foreach($value as $qp)
								$p[] = urlencode($param->name()) . '=' . urlencode($qp);
							$el = implode('&', $p);
						} else {
							$el = http_build_query( array( $param->name() => $value ) );
						}
					}
					else 
						$el = urlencode($param->name()) . '=' . urlencode($value);
					
					if(!empty($el)){
						$query[] = $el;
					}
					
					break;
					
				case 'header' : 
					
					if(preg_match('/^content-type$/i', $param->name())){
						$contentType = $value;
						$hasContentTypeHeader = true;
					}
					
					$headers[$param->name()] = $value;
					
					break;
					
				case 'path' : 
					$url = preg_replace('/\{'.$param->name().'\}/i', urlencode($value), $url);
					break;
					
				case 'body' : 
					
					$hasBody = true;
					$body = $value;
					
					break;
			}
			
			
			
		}
		
		
		$query = implode('&', $query);
		
		
		
		if(!empty($query)){
			$url .= strpos($url, '?')===false ? '?'.$query : '&'.$query;
		}
		
		
		// prepare the body
		if(!empty($formData)){
			// there is a form data
			
			// check if the consumes mime type are compatible
			$ph = array_intersect($consumes, array('application/x-www-form-urlencoded','multipart/form-data'));
			if(!empty($ph))
				$contentType = $ph[0];
			else
				$contentType = 'application/x-www-form-urlencoded'; // default
			
			
			if($hasFiles)
				$contentType = 'multipart/form-data'; // no choice !
			
			
			if($contentType === 'multipart/form-data'){
				// multipart form ...
				$boundary=md5(time());
				$contentType .= '; boundary=' . $boundary;
				$body = static::generateMultipartFormDataBody($formData, $boundary);
			}
			else {
				$body = implode('&', $formData);
			}
			
		}
		else if($hasBody){
			
			if(preg_match('/json/i', $contentType)){
				
				if(empty($body))
					$body = '{}';
				
				if(!is_string($body))
					$body = json_encode($body, JSON_PRETTY_PRINT);
				
			}
			else {
				throw new Exception("unable to encode the body data into '{$contentType}'");
			}
			
		}
		
		if(!$hasContentTypeHeader && $hasBody)
			$headers['Content-Type'] = $contentType;
		
		
		return array(
			'method' => strtoupper($method),
			'url' => $url,
			'headers' => $headers,
			'body' => $body
		);
		
		
	}
	
	
	
	private static function generateMultipartFormDataBody($data, $boundary){
		
		$body = '';
		$eol = "\r\n";
		
		foreach($data as $name => $d){
			
			$body .= '--' . $boundary . $eol;
			
			if($d['type']==='file'){
				$body .= 'Content-Disposition: form-data; name="'. $name .'"; filename="'. $d['filename'] .'"' . $eol;
				$body .= 'Content-Type: ' . $d['mime'] . $eol;
				$body .= $d['content'] . $eol;
			} else {
				$body .= 'Content-Disposition: form-data; name="'. $name .'"' . $eol . $eol;
				$body .= $d['content'] . $eol;
			}
		}
		
		$body .= "--" . $boundary . "--" . $eol . $eol;
		
		return $body;
	}

	
}

File: ething/swagger/Operation.py
# coding: utf-8
from future.utils import string_types

from ething.Helpers import reraise
import re
from .Parameter import instanciate as instanciate_parameter, StandardParameter
import sys
from ething.utils.mime import content_to_mime
import string
import random
import json
from future.utils import iteritems
try:
    from urllib.parse import quote, urlencode
except ImportError:
    from urllib import quote, urlencode


class Operation(object):

    def __init__(self, method, data, path):
        self.data = data
        self.__path = path
        self.__method = method
        self.__parameters = None

        if 'operationId' not in self.data:

            # generate a unique operation id

            opId = self.method + '_' + self.path.pathname

            opId = re.sub('[\s!@#$%^&*()_+=\[{\]};:<>|.\/?,\'"-]', '_', opId)
            opId = re.sub('((_){2,})', '_', opId)
            opId = re.sub('^_*', '', opId)
            opId = re.sub('_*$', '', opId)

            self.__name = opId

        else:
            self.__name = self.data.get('operationId')

    @property
    def root(self):
        return self.path.root

    @property
    def name(self):
        return self.__name

    @property
    def method(self):
        return self.__method

    @property
    def description(self):
        return self.data.get('description', '')

    @property
    def path(self):
        return self.__path

    @property
    def produces(self):
        return self.data.get('produces', self.root.produces)

    @property
    def consumes(self):
        c = self.data.get('consumes', self.root.consumes)
        if len(c) == 0:
            c.append('application/json')  # default to json
        return c

    @property
    def parameters(self):

        if self.__parameters is None:

            self.__parameters = []

            for paramobj in self.data.get('parameters', []):
                self.__parameters.append(instanciate_parameter(paramobj, self))

            self.__parameters += self.path.parameters

        return self.__parameters

    def createRequest(self, data=None, **options):

        # build the url
        schemes = self.data.get('schemes', self.root.schemes)
        if len(schemes) == 0:
            schemes = ['http']

        if 'scheme' in options and options['scheme'] in schemes:
            scheme = options['scheme']
        else:
            scheme = schemes[0]

        url = '%s://%s%s%s' % (scheme, self.root.host,
                               self.root.basePath, self.path.pathname)

        method = self.method
        query = []
        formData = {}
        headers = {}
        body = None
        hasFiles = 0
        hasContentTypeHeader = False
        hasBody = False

        consumes = self.consumes

        defaultContentType = 'application/json' if 'application/json' in consumes else consumes[0]
        contentType = options.get('requestContentType', defaultContentType)

        for param in self.parameters:

            hasValue = data is not None and param.name in data
            hasDefault = isinstance(
                param, StandardParameter) and param.hasDefault
            value = None

            if hasValue:
                value = data[param.name]
            else:
                if param.isRequired:
                    if hasDefault:
                        value = param.defaultValue
                    else:
                        raise Exception(
                            "missing required param: '%s'" % param.name)
                else:
                    continue

            # validate the incomming data
            try:
                v = param.validate(value)
            except Exception as e:
                reraise(type(e), type(e)("invalid param: '%s' : %s" %
                                         (param.name, str(e))), sys.exc_info()[2])
            else:
                value = v

            where = param.where

            if where == 'formData':

                if param.type == 'file':
                    
                    mime = content_to_mime(value, 'text/plain')

                    formData[param.name] = {
                        "type": "file",
                        "content": str(value),
                        "filename": 'file'+str(hasFiles),
                        "mime": mime
                    }

                    hasFiles += 1
                else:

                    formData[param.name] = {
                        "type": "text",
                        "content": str(value)
                    }

            elif where == 'query':

                if param.collectionFormat == 'multi':
                    for v in value:
                        query.append(quote(param.name.encode(
                            "utf-8"))+'='+quote(str(v).encode("utf-8")))
                else:
                    query.append(quote(param.name.encode("utf-8")) +
                                 '='+quote(str(value).encode("utf-8")))

            elif where == 'header':

                if param.name.lower() == 'content-type':
                    contentType = str(value)
                    hasContentTypeHeader = True

                headers[param.name] = value

            elif where == 'path':
                url = re.sub('\{%s\}' % param.name, urlencode(value), url)

            elif where == 'body':

                hasBody = True
                body = value

        query = '&'.join(query)

        if len(query) > 0:
            if '?' in url:
                url += '&' + query
            else:
                url += '?' + query

        # prepare the body
        if formData:
            # there is a form data
            # check if the consumes mime type are compatible

            if contentType != 'application/x-www-form-urlencoded' and contentType != 'multipart/form-data':
                # not compatible !
                # find a compatible contentType in the consumes list
                if 'application/x-www-form-urlencoded' in consumes:
                    contentType = 'application/x-www-form-urlencoded'
                elif 'multipart/form-data' in consumes:
                    contentType = 'multipart/form-data'
                else:
                    contentType = 'application/x-www-form-urlencoded'  # default

            if hasFiles:
                contentType = 'multipart/form-data'  # no choice !

            if contentType == 'multipart/form-data':
                # multipart form ...
                boundary = ''.join(random.choice(
                    string.ascii_lowercase + string.digits) for _ in range(10))
                contentType += '; boundary=' + boundary
                body = Operation.generateMultipartFormDataBody(
                    formData, boundary)

            else:

                p = []
                for k, v in iteritems(formData):
                    p.append(quote(k.encode("utf-8"))+'=' +
                             quote(str(v).encode("utf-8")))

                body = '&'.join(p)

        elif hasBody:

            if 'json' in contentType:

                if not body:
                    body = '{}'

                if not isinstance(body, string_types):
                    body = json.dumps(body, indent=4)

            elif isinstance(body, string_types):
                pass
            else:
                raise Exception(
                    "unable to encode the body data into '%s'" % contentType)

        if body and not hasContentTypeHeader:
            headers['Content-Type'] = contentType

        return {
            'method': method.upper(),
            'url': url,
            'headers': headers,
            'body': body
        }

    @staticmethod
    def generateMultipartFormDataBody(data, boundary):

        body = ''
        eol = "\r\n"

        for name, d in iteritems(data):

            body += '--' + boundary + eol

            if d['type'] == 'file':
                body += 'Content-Disposition: form-data; name="' + \
                    name + '"; filename="' + d['filename'] + '"' + eol
                body += 'Content-Type: ' + d['mime'] + eol
                body += d['content'] + eol
            else:
                body += 'Content-Disposition: form-data; name="' + name + '"' + eol + eol
                body += d['content'] + eol

        body += "--" + boundary + "--" + eol + eol

        return body
